### ISSUE TITLE
Wrong user selected as top tagger in certain cases

### ISSUE THREAD

**Issue by nonamethanks** [r888888888/danbooru#3587]
Seems there's some race condition that sometimes causes the wrong user to be selected as tagger.

Example: i only added one tag here:
https://danbooru.donmai.us/post_versions?search%5Bpost_id%5D=3067553
But I'm credited as the top tagger:
![](https://i.imgur.com/Ak7pEmP.png)

Pinging @evazion as he was the one who pointed out that it could be caused by a race condition since the timestamp of the edits is the same.

Edit: also those two tag removals in my first edit are too caused by a race condition, not the first time it happened. I'm sure I didn't remove them myself.

**Comment by nonamethanks** [comment]
Another example from a while ago of (possibly) the same race condition in action that i noticed some months ago: http://danbooru.donmai.us/post_versions?search%5Bpost_id%5D=2971584

**Comment by r888888888** [comment]
Short of implementing some sort of locking, maybe the best way to handle this (since I assume it isn't that common) is to fire an async job that runs a minute later to recalculate the top tagger.

**Comment by BrokenEagle** [comment]
One thing that also might need to be done is to choose the lower post version ID number when checking for whoever added a tag first.

https://danbooru.donmai.us/post_versions.xml?search%5Bpost_id%5D=3087760

As can be seen with above, due to some bug, both users have virtually the same sets of added tags. However, they have the same timestamp, but for whatever reason , the system is choosing the next tagger instead of the original uploader.  Using the post version ID number instead of the ``updated_at`` timestamp should fix that I believe.

**Comment by evazion** [comment]
It doesn't need locking. Versions just need to be ordered by ID instead of by timestamp.

Locking wouldn't do anything in any case, since archives processes versions one at a time. As long as there is still only one queue worker, versions are never processed in parallel.

**Comment by evazion** [comment]
No longer relevant now that top tagger is gone.
